{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ce6d7697",
   "metadata": {},
   "source": [
    "# Annotating the Chromsome 4 of *Babesia duncani*\n",
    "\n",
    "Originally, we had intended to provide an exercise for annotating a complete small genome. However, the computational resources available on AWS for this workshop do not permit completion of the task within one afternoon. Therefore, we provide the annotation of Chromosome 4 of *Babesia duncani* as a hands-on exercise for practicing your skills. \n",
    "\n",
    "\n",
    "*Babesia duncani* is a protozoan parasite that infects red blood cells in humans. It is primarily transmitted through tick bites. *Babesia duncani* is known to cause a malaria-like illness called babesiosis, which can lead to symptoms such as fever, fatigue, muscle aches, and anemia. The [genome](https://www.ncbi.nlm.nih.gov/data-hub/taxonomy/323732/) has a size of 10.4 Mb. We selected Chromsome 4 from this genome, which is about 1 Mb in size. \n",
    "\n",
    "**Warning:* None of the self-training genome annotation tools (BRAKER1, BRAKER2, BRAKER3, GALBA) are particularly suitable for annotating partial genomes. We strongly advise against doing this in practice. Please always use genomes as complete as possible. The only reason why we resort to a chunk of a genome, here, are limited computational and time resources during this workshop.\n",
    "  \n",
    "Your task is today to annotate this partial genome with several methods: BRAKER3, BRAKER2, BRAKER1, GALBA, and possibly TSEBRA combinations of the beforementioned, and (a) produce a high quality annotation in a fully automated way, and (b) find out which method produces best results.\n",
    "\n",
    "Beware of the secret cookie drawer! We keep solutions in the secret cookie drawer. If you touch the contents of the secret cookie drawer, you'll spoil your learning process! Don't do it! (Unless it's absolutely necessary.)\n",
    "\n",
    "\n",
    "## Data\n",
    "\n",
    "The genome is available at [/home/genomics/workshop_materials/genome_annotation/genome/genome.fa](/home/genomics/workshop_materials/genome_annotation/genome/genome.fa). We will be using the repeat masking provided by NCBI.\n",
    "\n",
    "We downloaded and aligned the library SRR18907291 from [Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra). (Paired-end Illumina library, Hisat2 aligner, converted and sorted with Samtools, selected aligned reads with Samtools to reduce bam file size.) It is not feasible that we process this data set during the workhop. We provide the file [/home/genomics/workshop_materials/genome_annotation/sra/SRR18907291.s.mapped.bam](/home/genomics/workshop_materials/genome_annotation/sra/SRR18907291.s.mapped.bam).\n",
    "\n",
    "In terms of OrthoDB taxonomy, the Alveolata partition is \"the closest\" for this species, available at [/home/genomics/workshop_materials/genome_annotation/orthodb/Alveolata.fa](/home/genomics/workshop_materials/genome_annotation/orthodb/Alveolata.fa). This is a rather \"small\" partition, and a priori, we do not know whether it will be sufficient for running BRAKER3 or BRAKER2 if used as the only protein input. We recommend combining it with proteins of other *Babesia* species (see below).\n",
    "\n",
    "Proteomes of other *Babesia* species were downloaded from NCBI, they were concatenated, and the fasta headers have been replaced by short, simple, and unique headers. The resulting file to be used with GALBA is available at [/home/genomics/workshop_materials/genome_annotation/proteins/proteins.fa](/home/genomics/workshop_materials/genome_annotation/proteins/proteins.fa). \n",
    "\n",
    "## What to expect?\n",
    "\n",
    "This genome has no complete reference annotation, yet. Looking at other *Babesia* genomes, we exepect 3,000 to 5,000 proteins in the full genome. Chromosome 4 makes up only 10% of the total genome, i.e. we possibly expect 300 to 500 proteins to be predicted in this genome part.\n",
    "\n",
    "We applied BUSCO to the genome as follows:\n",
    "\n",
    "```\n",
    "busco -m genome -i genome.fa -o busco_genome -l alveolata_odb10\n",
    "```\n",
    "\n",
    "and obtained this output:\n",
    "\n",
    "```\n",
    "C:17.5%[S:17.5%,D:0.0%],F:1.2%,M:81.3%,n:171\n",
    "```\n",
    "\n",
    "The BUSCO core gene set for Alveolata is relatively small (only 171 genes). However, we expect similar sensitivity scores with a well-performing tool for structural genome annotation.\n",
    "\n",
    "## Group assignment\n",
    "\n",
    "Runtime is a serious issue. Therefore, you are split into 4 different groups, and you will at first only complete the task of the group that you were assigned to. Execute the following code cell once to see which group are in:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d96887ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "print(\"You are in group \", random.randint(1,5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11603735",
   "metadata": {},
   "source": [
    "For all tasks, there is a possibility of failure. Should your task fail (e.g. because not sufficient evidence was provided), pick a new group by re-executing the above code cell.\n",
    "\n",
    "## Questions to all groups (with respect to the assigned task)\n",
    "\n",
    "1. After executing the annotation task assigned to you, apply BUSCO to estimate proteome completentess (use [/home/genomics/workshop_materials/genome_annotation/alveolata_odb10](/home/genomics/workshop_materials/genome_annotation/alveolata_odb10)). What are the BUSCO scores?\n",
    "\n",
    "2. How many genes and how many transcripts are predicted?\n",
    "\n",
    "3. How is the mono:mult ratio? \n",
    "\n",
    "4. What is the median number of exons per transcript? \n",
    "\n",
    "5. What is the largest number of exons in a transcript?\n",
    "\n",
    "Fill in the results for your task(s) at our [Google Results Sheet](https://docs.google.com/spreadsheets/d/1WT3LGMrvWS-ai3xRjFLVvFMSUlLTfqMFnYISbQES99I/edit?usp=sharing) in the grey fields. If no results can be obtained, enter \"NA\".\n",
    "\n",
    "## Common pitfalls\n",
    "\n",
    "   * **Warning** BRAKER may warn you about a low number of introns according to evidence. This can be caused either by a species not having many introns, or by a lack of evidence. In our particular case, it will be caused by using only 10% of a complete genome, and by little evidence.\n",
    "   \n",
    "   * **Warning** BRAKER or GALBA may warn you that the number of training genes is very low. In our example, we are using only a partial genome, and for today, we will ignore this warning. However, in a real-life scenario with a complete genome, you should be cautious when seeing such a warning. It's not advisable to train a gene finder with less than 200 genes. Having more than 200 training genes may or may not give good results. A large number of training genes (no warning shown) often leads to good results.\n",
    "   \n",
    "   * Sometimes, **a step in the pipeline may fail** when processing your specific dataset. BRAKER usually reports that the most common problem is a missing or expired file called ~/.gm_key. However, during the workshop, we can assure you that the key file is not expired or missing! To find the actual reason for the failure, you can check the error file of the last command executed by BRAKER. You can find that command in the braker.log file in the working directory. Most often, the reason for the failure is missing data. In this workshop, we won't add any additional data to solve this problem as it would increase the runtime. However, in real-life scenarios, you may be able to overcome the issue by adding more RNA-Seq data or more protein data.\n",
    "\n",
    "## Task for group 1: Apply GALBA\n",
    "\n",
    "Use the following code cell to implement the application of GALBA, and execute GALBA. Use the genome file and the proteins.fa file mentioned, above. Apply `--skipOptimize`. (Expected runtime: ~10 minutes.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dfea1b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to analyze results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d16bccba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to analyze results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b78c11f1",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Secret cookie drawer. Don't touch it! Don't.... !!!</b></summary>\n",
    "Here is how to run GALBA:\n",
    "    \n",
    "    \n",
    "```\n",
    "%%script bash\n",
    "T=8 # adjust to number of threads that you booted with\n",
    "\n",
    "# delete output from a possible previous run if it exists\n",
    "if [ -d GALBA_babesia ]\n",
    "then\n",
    "    rm -rf GALBA_babesia\n",
    "fi\n",
    "\n",
    "time galba.pl --workingdir=GALBA_babesia --genome=/home/genomics/workshop_materials/genome_annotation/genome/genome.fa \\\n",
    "    --prot_seq=/home/genomics/workshop_materials/genome_annotation/proteins/proteins.fa \\\n",
    "    --AUGUSTUS_BIN_PATH=/usr/bin/ --AUGUSTUS_SCRIPTS_PATH=/usr/share/augustus/scripts/ --threads ${T} \\\n",
    "    --skipOptimize\n",
    "```\n",
    "\n",
    "Here's how to analyze results:\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with\n",
    "\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "cd GALBA_babesia\n",
    "# create links if not already present\n",
    "if [ ! -d busco_downloads ]\n",
    "then\n",
    "    mkdir busco_downloads\n",
    "    cd busco_downloads\n",
    "    if [ ! -L alveolata_odb10 ]\n",
    "    then\n",
    "        ln -s /home/genomics/workshop_materials/genome_annotation/busco/alveolata_odb10 alveolata_odb10\n",
    "    fi\n",
    "    cd ..\n",
    "fi\n",
    "\n",
    "# delete old output if existing\n",
    "if [ -d busco_galba_babesia ]\n",
    "then\n",
    "    rm -r busco_galba_babesia\n",
    "fi\n",
    "\n",
    "# run BUSCO\n",
    "busco -m proteins -i augustus.hints.aa -o busco_galba_babesia \\\n",
    "        -l alveolata_odb10 -c ${T} &> busco_galba_babesia.log\n",
    "\n",
    "# Count number of transcripts by counting FASTA headers\n",
    "echo \"Counting number of protein sequences = transcripts\"\n",
    "grep -c \">\" augustus.hints.aa\n",
    "\n",
    "# Number of genes, and descriptive statistics\n",
    "echo \"Computing some descriptive statistics for GALBA:\"\n",
    "analyze_exons.py -f augustus.hints.gtf\n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "## Task for group 2: Apply BRAKER3\n",
    "\n",
    "Use the following code cell to implement the application of BRAKER3, and execute BRAKER3. Use the genome file, the OrthoDB Metazoa file, the proteins.fa file, and the Merged.bam file mentioned, above. Apply `--skipOptimize`. Do not apply `--gm_max_intergenic 10000`. (Expected runtime: ~10 minutes.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e76205db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to analyze results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c48050e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to analyze results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63d27777",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Secret cookie drawer. Don't touch it! Don't.... !!!</b></summary>\n",
    "Here is how to run BRAKER:\n",
    "    \n",
    "    \n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with, takes ~30 minutes with 4 threads\n",
    "\n",
    "# delete output from a possible previous run if it exists\n",
    "if [ -d BRAKER3_babesia ]\n",
    "then\n",
    "    rm -rf BRAKER3_babesia\n",
    "fi\n",
    "\n",
    "ORTHODB=/home/genomics/workshop_materials/genome_annotation/orthodb/Alveolata.fa # adjust to suitable clade\n",
    "\n",
    "# run BRAKER3\n",
    "time braker.pl --workingdir=BRAKER3_babesia --genome=/home/genomics/workshop_materials/genome_annotation/genome/genome.fa \\\n",
    "    --bam=/home/genomics/workshop_materials/genome_annotation/sra/SRR18907291.s.mapped.bam \\\n",
    "    --prot_seq=${ORTHODB},/home/genomics/workshop_materials/genome_annotation/proteins/proteins.fa \\\n",
    "    --AUGUSTUS_BIN_PATH=/usr/bin/ \\\n",
    "    --AUGUSTUS_SCRIPTS_PATH=/usr/share/augustus/scripts/ --threads ${T} \\\n",
    "    --skipOptimize\n",
    "```\n",
    "\n",
    "Here's how to analyze results (in theory, actually, BRAKER3 dies on this dataset):\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with\n",
    "\n",
    "# CANNOT BE EXECUTED, BRAKER3 DIES!\n",
    "\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "cd BRAKER3_babesia\n",
    "# create links if not already present\n",
    "if [ ! -d busco_downloads ]\n",
    "then\n",
    "    mkdir busco_downloads\n",
    "    cd busco_downloads\n",
    "    if [ ! -L alveolata_odb10 ]\n",
    "    then\n",
    "        ln -s /home/genomics/workshop_materials/genome_annotation/busco/alveolata_odb10 alveolata_odb10\n",
    "    fi\n",
    "    cd ..\n",
    "fi\n",
    "\n",
    "# delete old output if existing\n",
    "if [ -d busco_braker3_babesia ]\n",
    "then\n",
    "    rm -r busco_braker3_babesia\n",
    "fi\n",
    "\n",
    "# run BUSCO\n",
    "busco -m proteins -i braker.aa -o busco_braker3_babesia \\\n",
    "        -l alveolata_odb10 -c ${T} &> busco_braker3_babesia.log\n",
    "\n",
    "# Count number of transcripts by counting FASTA headers\n",
    "echo \"Counting number of protein sequences = transcripts\"\n",
    "grep -c \">\" braker.aa\n",
    "\n",
    "# Number of genes, and descriptive statistics\n",
    "echo \"Computing some descriptive statistics for BRAKER3:\"\n",
    "analyze_exons.py -f braker.gtf\n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "## Task for group 3: Apply BRAKER2\n",
    "\n",
    "Use the following code cell to implement the application of BRAKER2, and execute BRAKER2. Use the genome file, the OrthoDB Metazoa file and the proteins.fa file mentioned, above. Apply `--skipOptimize`. Do not apply `--gm_max_intergenic 10000`. (Expected runtime: ~55 minutes.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee296dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to execute gene finder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94c14622",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to analyze results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeafea05",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Secret cookie drawer. Don't touch it! Don't.... !!!</b></summary>\n",
    "Here is how to run BRAKER:\n",
    "    \n",
    "    \n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with, takes ~30 minutes with 4 threads\n",
    "\n",
    "# delete output from a possible previous run if it exists\n",
    "if [ -d BRAKER2_babesia ]\n",
    "then\n",
    "    rm -rf BRAKER2_babesia\n",
    "fi\n",
    "\n",
    "ORTHODB=/home/genomics/workshop_materials/genome_annotation/orthodb/Alveolata.fa # adjust to suitable clade\n",
    "\n",
    "# run BRAKER2\n",
    "time braker.pl --workingdir=BRAKER2_babesia --genome=/home/genomics/workshop_materials/genome_annotation/genome/genome.fa \\\n",
    "    --prot_seq=${ORTHODB},/home/genomics/workshop_materials/genome_annotation/proteins/proteins.fa \\\n",
    "    --AUGUSTUS_BIN_PATH=/usr/bin/ \\\n",
    "    --AUGUSTUS_SCRIPTS_PATH=/usr/share/augustus/scripts/ --threads ${T} \\\n",
    "    --skipOptimize\n",
    "```\n",
    "\n",
    "Here's how to analyze results:\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with\n",
    "\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "cd BRAKER2_babesia\n",
    "# create links if not already present\n",
    "if [ ! -d busco_downloads ]\n",
    "then\n",
    "    mkdir busco_downloads\n",
    "    cd busco_downloads\n",
    "    if [ ! -L alveolata_odb10 ]\n",
    "    then\n",
    "        ln -s /home/genomics/workshop_materials/genome_annotation/busco/alveolata_odb10 alveolata_odb10\n",
    "    fi\n",
    "    cd ..\n",
    "fi\n",
    "\n",
    "# delete old output if existing\n",
    "if [ -d busco_braker2_babesia ]\n",
    "then\n",
    "    rm -r busco_braker2_babesia\n",
    "fi\n",
    "\n",
    "# run BUSCO\n",
    "busco -m proteins -i braker.aa -o busco_braker2_babesia \\\n",
    "        -l alveolata_odb10 -c ${T} &> busco_braker2_babesia.log\n",
    "\n",
    "# Count number of transcripts by counting FASTA headers\n",
    "echo \"Counting number of protein sequences = transcripts\"\n",
    "grep -c \">\" braker.aa\n",
    "\n",
    "# Number of genes, and descriptive statistics\n",
    "echo \"Computing some descriptive statistics for GALBA:\"\n",
    "analyze_exons.py -f braker.gtf\n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "## Task for group 4: Apply BRAKER1\n",
    "\n",
    "Use the following code cell to implement the application of BRAKER1, and execute BRAKER1. Use the genome file and the Merged.bam file mentioned, above. Apply `--skipOptimize`. Do not apply `--gm_max_intergenic 10000`. (Expected runtime: ~10 minutes.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02f4af19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to execute gene finder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d2d314d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to analyze results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0828b37",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Secret cookie drawer. Don't touch it! Don't.... !!!</b></summary>\n",
    "Here is how to run BRAKER:\n",
    "    \n",
    "    \n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with, takes ~30 minutes with 4 threads\n",
    "\n",
    "# delete output from a possible previous run if it exists\n",
    "if [ -d BRAKER1_babesia ]\n",
    "then\n",
    "    rm -rf BRAKER1_babesia\n",
    "fi\n",
    "\n",
    "# run BRAKER1\n",
    "time braker.pl --workingdir=BRAKER1_babesia \\\n",
    "    --genome=/home/genomics/workshop_materials/genome_annotation/genome/genome.fa \\\n",
    "    --bam=/home/genomics/workshop_materials/genome_annotation/sra/SRR18907291.s.mapped.bam \\\n",
    "    --AUGUSTUS_BIN_PATH=/usr/bin/ \\\n",
    "    --AUGUSTUS_SCRIPTS_PATH=/usr/share/augustus/scripts/ --threads ${T} \\\n",
    "    --skipOptimize\n",
    "```\n",
    "\n",
    "Here's how to analyze results:\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with\n",
    "\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "cd BRAKER1_babesia\n",
    "# create links if not already present\n",
    "if [ ! -d busco_downloads ]\n",
    "then\n",
    "    mkdir busco_downloads\n",
    "    cd busco_downloads\n",
    "    if [ ! -L alveolata_odb10 ]\n",
    "    then\n",
    "        ln -s /home/genomics/workshop_materials/genome_annotation/busco/alveolata_odb10 alveolata_odb10\n",
    "    fi\n",
    "    cd ../\n",
    "fi\n",
    "\n",
    "# delete old output if existing\n",
    "if [ -d busco_braker1_babesia ]\n",
    "then\n",
    "    rm -r busco_braker1_babesia\n",
    "fi\n",
    "\n",
    "# run BUSCO\n",
    "busco -m proteins -i braker.aa -o busco_braker1_babesia \\\n",
    "        -l alveolata_odb10 -c ${T} &> busco_braker1_babesia.log\n",
    "\n",
    "# Count number of transcripts by counting FASTA headers\n",
    "echo \"Counting number of protein sequences = transcripts\"\n",
    "grep -c \">\" braker.aa\n",
    "\n",
    "# Number of genes, and descriptive statistics\n",
    "echo \"Computing some descriptive statistics:\"\n",
    "analyze_exons.py -f braker.gtf\n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "## Do we have a sensitivity problem?\n",
    "\n",
    "When comparing the obtained BUSCO scores, you may wonder why we are missing BUSCOs that were detected on genome level. \n",
    "\n",
    "One possible reason is that MetaEuk, the gene finder employed to detect BUSCOs on genome level, may find BUSCOs in genomic regions that have been fully repeat masked. AUGUSTUS cannot find these genes (applies to both BRAKER and GALBA). \n",
    "\n",
    "Another reason, particularly in BRAKER, may be the lack of evidence, and poor training (remember, we are using only a small proportion of a small genomes, and we skipped an optimization step for AUGUSTUS training). If you have time, play with TSEBRA to generate a possibly superior gene set. Consider combining the gene sets of BRAKER1 and BRAKER2, or even of BRAKER1, BRAKER2, and BRAKER3. Keep in mind that we learned how to enforce a gene set even in lack of evidence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f29d041d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to run TSEBRA with BRAKER1 and BRAKER2 if you have these results and have time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60fd5e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert code to run TSEBRA with BRAKER1, BRAKER2, and GALBA if you have these results and have time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8003dec9",
   "metadata": {},
   "source": [
    "<details>\n",
    "  <summary><b>Secret cookie drawer. Don't touch it! Don't.... !!!</b></summary>\n",
    "Here is how to run TSEBEA with BRAKER1 and BRAKER2:\n",
    "    \n",
    "    \n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with\n",
    "\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "if [ -d TSEBRA_B1_B2 ]\n",
    "then\n",
    "    rm -r TSEBRA_B1_B2\n",
    "fi\n",
    "\n",
    "mkdir TSEBRA_B1_B2\n",
    "cd TSEBRA_B1_B2\n",
    "tsebra.py -k ../BRAKER2_babesia/braker.gtf -g ../BRAKER1_babesia/braker.gtf \\\n",
    "    -e ../BRAKER2_babesia/hintsfile.gff,../BRAKER1_babesia/hintsfile.gff -o tsebra.gtf 2> tsebra.log\n",
    "\n",
    "getAnnoFastaFromJoingenes.py -g /home/genomics/workshop_materials/genome_annotation/genome/genome.fa \\\n",
    "    -o tsebra -f tsebra.gtf\n",
    "\n",
    "# create links if not already present\n",
    "if [ ! -d busco_downloads ]\n",
    "then\n",
    "    mkdir busco_downloads\n",
    "    cd busco_downloads\n",
    "    if [ ! -L alveolata_odb10 ]\n",
    "    then\n",
    "        ln -s /home/genomics/workshop_materials/genome_annotation/busco/alveolata_odb10 alveolata_odb10\n",
    "    fi\n",
    "    cd ../\n",
    "fi\n",
    "\n",
    "# delete old output if existing\n",
    "if [ -d busco_tsebra_b1_b2 ]\n",
    "then\n",
    "    rm -r busco_tsebra_b1_b2\n",
    "fi\n",
    "\n",
    "# run BUSCO\n",
    "busco -m proteins -i tsebra.aa -o busco_tsebra_b1_b2 \\\n",
    "        -l alveolata_odb10 -c ${T} &> busco_tsebra_b1_b2.log\n",
    "\n",
    "# Count number of transcripts by counting FASTA headers\n",
    "echo \"Counting number of protein sequences = transcripts\"\n",
    "grep -c \">\" tsebra.aa\n",
    "\n",
    "# Number of genes, and descriptive statistics\n",
    "echo \"Computing some descriptive statistics for TSEBRA:\"\n",
    "analyze_exons.py -f tsebra.gtf\n",
    "```\n",
    "\n",
    "Here's how run it with GALBA in addition\n",
    "\n",
    "```\n",
    "%%script bash\n",
    "\n",
    "T=8 # adjust to number of threads that you booted with\n",
    "\n",
    "source conda_init\n",
    "conda activate busco_env\n",
    "\n",
    "if [ -d TSEBRA_B1_B2_G ]\n",
    "then\n",
    "    rm -r TSEBRA_B1_B2_G\n",
    "fi\n",
    "\n",
    "mkdir TSEBRA_B1_B2_G\n",
    "cd TSEBRA_B1_B2_G\n",
    "tsebra.py -k ../BRAKER2_babesia/braker.gtf -g ../BRAKER1_babesia/braker.gtf,../GALBA_babesia/augustus.hints.gtf \\\n",
    "    -e ../BRAKER2_babesia/hintsfile.gff,../BRAKER1_babesia/hintsfile.gff,../GALBA_babesia/hintsfile.gff -o tsebra.gtf 2> tsebra.log\n",
    "\n",
    "getAnnoFastaFromJoingenes.py -g /home/genomics/workshop_materials/genome_annotation/genome/genome.fa \\\n",
    "    -o tsebra -f tsebra.gtf\n",
    "\n",
    "# create links if not already present\n",
    "if [ ! -d busco_downloads ]\n",
    "then\n",
    "    mkdir busco_downloads\n",
    "    cd busco_downloads\n",
    "    if [ ! -L alveolata_odb10 ]\n",
    "    then\n",
    "        ln -s /home/genomics/workshop_materials/genome_annotation/busco/alveolata_odb10 alveolata_odb10\n",
    "    fi\n",
    "    cd ../\n",
    "fi\n",
    "\n",
    "# delete old output if existing\n",
    "if [ -d busco_tsebra_b1_b2_g ]\n",
    "then\n",
    "    rm -r busco_tsebra_b1_b2_g\n",
    "fi\n",
    "\n",
    "# run BUSCO\n",
    "busco -m proteins -i tsebra.aa -o busco_tsebra_b1_b2_g \\\n",
    "        -l alveolata_odb10 -c ${T} &> busco_tsebra_b1_b2_g.log\n",
    "\n",
    "# Count number of transcripts by counting FASTA headers\n",
    "echo \"Counting number of protein sequences = transcripts\"\n",
    "grep -c \">\" tsebra.aa\n",
    "\n",
    "# Number of genes, and descriptive statistics\n",
    "echo \"Computing some descriptive statistics for TSEBRA:\"\n",
    "analyze_exons.py -f tsebra.gtf\n",
    "```\n",
    "\n",
    "</details>\n",
    "\n",
    "We are close to the end of our genome annotation journey!\n",
    "\n",
    "<details>\n",
    "  <summary><b>Out of time but want to see results? Click here!</b></summary>\n",
    "<a href=\"https://docs.google.com/spreadsheets/d/14zgTHt8ZPnGzKffJ7x3e_HWwqoOF6hH9GfbLS6a1I-Q/edit?usp=sharing\">Goolge Sheet with Results</a>\n",
    "</details>\n",
    "\n",
    "### The End"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b85db7f3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
